code/datums/components/bayonet_attachable.dm

```diff
@@ -0,0 +1,212 @@
+	/// Formatted as parent_base_state-[bayonet_icon_state-state]
+	var/bayonet_icon_state
+	/// If passed, we will use a specific overlay instead of using the knife itself
+	/// The state to take from the bayonet overlay icon if supplied.
+	var/bayonet_overlay
+	/// This is the icon file it grabs the overlay from.
+	var/bayonet_overlay_icon
+	/// Offsets for the bayonet overlay
+	var/offset_x = 0
+	var/offset_y = 0
+	/// If this component allows sawing off the parent gun/should be deleted when the parent gun is sawn off
+	var/allow_sawnoff = FALSE
+
+	// Internal vars
+	/// Currently attached bayonet
+	var/obj/item/knife/bayonet
+	/// Static typecache of all knives that can become bayonets
+	var/static/list/valid_bayonets = typecacheof(list(/obj/item/knife/combat))
+
+/datum/component/bayonet_attachable/Initialize(
+	obj/item/knife/starting_bayonet,
+	offset_x = 0,
+	offset_y = 0,
+	removable = TRUE,
+	bayonet_icon_state = null,
+	bayonet_overlay = "bayonet",
+	bayonet_overlay_icon = 'icons/obj/weapons/guns/bayonets.dmi',
+	allow_sawnoff = FALSE
+)
+
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.removable = removable
+	src.bayonet_icon_state = bayonet_icon_state
+	src.bayonet_overlay = bayonet_overlay
+	src.bayonet_overlay_icon = bayonet_overlay_icon
+	src.offset_x = offset_x
+	src.offset_y = offset_y
+	src.allow_sawnoff = allow_sawnoff
+
+	if (istype(starting_bayonet))
+		add_bayonet(starting_bayonet)
+
+/datum/component/bayonet_attachable/Destroy(force)
+	if(bayonet)
+		remove_bayonet()
+	return ..()
+
+/datum/component/bayonet_attachable/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_OBJ_DECONSTRUCT, PROC_REF(on_parent_deconstructed))
+	RegisterSignal(parent, COMSIG_ATOM_EXITED, PROC_REF(on_item_exit))
+	RegisterSignal(parent, COMSIG_ATOM_TOOL_ACT(TOOL_SCREWDRIVER), PROC_REF(on_screwdriver))
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_ICON_STATE, PROC_REF(on_update_icon_state))
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, PROC_REF(on_update_overlays))
+	RegisterSignal(parent, COMSIG_ATOM_ATTACKBY, PROC_REF(on_attackby))
+	RegisterSignal(parent, COMSIG_ATOM_EXAMINE, PROC_REF(on_examine))
+	RegisterSignal(parent, COMSIG_QDELETING, PROC_REF(on_parent_deleted))
+	RegisterSignal(parent, COMSIG_ITEM_PRE_ATTACK, PROC_REF(on_pre_attack))
+	RegisterSignal(parent, COMSIG_GUN_BEING_SAWNOFF, PROC_REF(on_being_sawnoff))
+	RegisterSignal(parent, COMSIG_GUN_SAWN_OFF, PROC_REF(on_sawn_off))
+
+/datum/component/bayonet_attachable/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_OBJ_DECONSTRUCT,
+		COMSIG_ATOM_EXITED,
+		COMSIG_ATOM_TOOL_ACT(TOOL_SCREWDRIVER),
+		COMSIG_ATOM_UPDATE_ICON_STATE,
+		COMSIG_ATOM_UPDATE_OVERLAYS,
+		COMSIG_ATOM_ATTACKBY,
+		COMSIG_ATOM_EXAMINE,
+		COMSIG_QDELETING,
+		COMSIG_ITEM_PRE_ATTACK,
+		COMSIG_GUN_BEING_SAWNOFF,
+		COMSIG_GUN_SAWN_OFF,
+	))
+
+/datum/component/bayonet_attachable/proc/on_examine(obj/item/source, mob/examiner, list/examine_list)
+	SIGNAL_HANDLER
+
+	if(!bayonet)
```

Comment:
```suggestion
	if(isnull(bayonet))
```
We prefer specificity. Or at least San does and always told me to do this while reviewing my PRs, so I am passing it on.